File: tools/server_observer_cpp/src/observation_api.cpp
#include "observation_api.hpp"
#include <httplib.h>
#include <iostream>
#include <chrono>
#include <sstream>
#include <iomanip>
#include <openssl/sha.h>

static const int SUPPORTED_CLIENT_VERSION = 207;
static const int MAX_RETRIES = 3;

static std::string sha1_hex(const std::string& input) {
    unsigned char hash[SHA_DIGEST_LENGTH];
    SHA1(reinterpret_cast<const unsigned char*>(input.c_str()), input.size(), hash);
    
    std::stringstream ss;
    for (int i = 0; i < SHA_DIGEST_LENGTH; i++) {
        ss << std::hex << std::setw(2) << std::setfill('0') << static_cast<int>(hash[i]);
    }
    return ss.str();
}

static int64_t current_time_ms() {
    auto now = std::chrono::system_clock::now();
    auto ms = std::chrono::duration_cast<std::chrono::milliseconds>(now.time_since_epoch());
    return ms.count();
}

ObservationApi::ObservationApi(const json& headers,
                             const json& cookies,
                             const json& proxy,
                             const AuthDetails& auth_details,
                             int game_id,
                             const std::string& game_server_address,
                             int client_version)
    : game_id_(game_id)
    , player_id_(0)
    , auth_(auth_details)
    , request_id_(0)
    , client_version_(client_version)
    , game_server_address_(game_server_address)
    , headers_(headers)
    , cookies_(cookies)
    , proxy_(proxy)
{
}

ObservationApi::~ObservationApi() {
}

json ObservationApi::make_game_server_request(const json& parameters) {
    int attempt = 0;
    
    while (true) {
        // Parse game server address
        std::string host, path;
        int port = 443;
        bool use_ssl = true;
        
        if (game_server_address_.find("https://") == 0) {
            host = game_server_address_.substr(8);
            use_ssl = true;
            port = 443;
        } else if (game_server_address_.find("http://") == 0) {
            host = game_server_address_.substr(7);
            use_ssl = false;
            port = 80;
        } else {
            host = game_server_address_;
        }
        
        // Extract path if present
        size_t slash_pos = host.find('/');
        if (slash_pos != std::string::npos) {
            path = host.substr(slash_pos);
            host = host.substr(0, slash_pos);
        } else {
            path = "/";
        }
        
        // Create HTTP client
        httplib::Client cli(host.c_str(), port);
        cli.set_follow_location(true);
        
        if (use_ssl) {
            cli.enable_server_certificate_verification(false);
        }
        
        // Build request headers
        httplib::Headers req_headers = {
            {"Accept", "text/plain, */*; q=0.01"},
            {"Content-Type", "application/x-www-form-urlencoded; charset=UTF-8"},
            {"Accept-Encoding", "gzip, deflate, br"}
        };
        
        // Add custom headers
        if (headers_.is_object()) {
            for (auto& [key, value] : headers_.items()) {
                if (value.is_string()) {
                    req_headers.insert({key, value.get<std::string>()});
                }
            }
        }
        
        // Build payload
        std::string hash_input = "undefined" + std::to_string(current_time_ms());
        std::string hash_hex = sha1_hex(hash_input);
        
        json payload = {
            {"requestID", request_id_},
            {"language", "en"},
            {"version", client_version_},
            {"tstamp", std::to_string(auth_.auth_tstamp)},
            {"client", "con-client"},
            {"hash", hash_hex},
            {"sessionTstamp", 0},
            {"gameID", std::to_string(game_id_)},
            {"playerID", player_id_},
            {"siteUserID", std::to_string(auth_.user_id)},
            {"adminLevel", nullptr},
            {"rights", auth_.rights},
            {"userAuth", auth_.auth},
            {"lastCallDuration", 0}
        };
        
        // Merge parameters into payload
        for (auto& [key, value] : parameters.items()) {
            payload[key] = value;
        }
        
        request_id_++;
        
        // Send request
        std::string body = payload.dump();
        auto res = cli.Post(path.c_str(), req_headers, body, "application/json");
        
        if (!res) {
            throw std::runtime_error("HTTP request failed");
        }
        
        if (res->status != 200) {
            throw std::runtime_error("HTTP status: " + std::to_string(res->status));
        }
        
        // Parse response
        json response_json;
        try {
            response_json = json::parse(res->body);
        } catch (const std::exception& e) {
            throw std::runtime_error("Failed to parse response: " + std::string(e.what()));
        }
        
        // Update server time
        if (response_json.contains("result")) {
            auto& result = response_json["result"];
            if (result.is_object() && result.contains("timeStamp")) {
                update_server_time(result["timeStamp"].get<int64_t>());
            } else {
                update_server_time(0);
            }
        }
        
        // Handle errors
        if (response_json.contains("result") && response_json["result"].is_object()) {
            auto& result = response_json["result"];
            
            if (result.contains("@c")) {
                std::string err_class = result["@c"].get<std::string>();
                
                if (err_class == "ultshared.UltAuthentificationException") {
                    throw std::runtime_error("Authentication failed");
                }
                
                if (err_class == "ultshared.rpc.UltSwitchServerException") {
                    // Update server address
                    if (result.contains("newHostName")) {
                        std::string new_server = "https://" + result["newHostName"].get<std::string>();
                        std::cout << "Switching game server to " << new_server << std::endl;
                        game_server_address_ = new_server;
                    }
                    
                    // Retry
                    if (attempt >= MAX_RETRIES) {
                        throw std::runtime_error("Exceeded retries after server switch suggestion");
                    }
                    attempt++;
                    std::cout << "Retrying after UltSwitchServerException (attempt " 
                             << attempt << "/" << MAX_RETRIES << ")" << std::endl;
                    continue;
                }
            }
        }
        
        return response_json;
    }
}

void ObservationApi::update_server_time(int64_t t_stamp_now) {
    // In C++, we don't need to track server time offset for observation
    // This is a simplified implementation
}

json ObservationApi::request_game_state(std::map<int, std::string>& state_ids,
                                        std::map<int, int>& time_stamps) {
    bool include_state_meta = !state_ids.empty() && !time_stamps.empty();
    
    // Build state_ids and time_stamps as JSON objects
    json state_ids_json = json::object();
    json time_stamps_json = json::object();
    
    if (include_state_meta) {
        for (const auto& [key, value] : state_ids) {
            state_ids_json[std::to_string(key)] = value;
        }
        for (const auto& [key, value] : time_stamps) {
            time_stamps_json[std::to_string(key)] = value;
        }
    }
    
    // Build action
    json action = {
        {"@c", "ultshared.action.UltUpdateGameStateAction"},
        {"stateType", 0},
        {"stateID", "0"},
        {"addStateIDsOnSent", include_state_meta}
    };
    
    if (include_state_meta) {
        action["stateIDs"] = {
            {"@c", "java.util.HashMap"},
            {"@m", state_ids_json}
        };
        action["timeStamps"] = {
            {"@c", "java.util.HashMap"},
            {"@m", time_stamps_json}
        };
    }
    
    action["actions"] = {
        {"@c", "java.util.LinkedList"}
    };
    
    json response = make_game_server_request(action);
    
    if (!extract_state_metadata(response, state_ids, time_stamps)) {
        throw std::runtime_error("Game state extraction failed");
    }
    
    return response;
}

bool ObservationApi::extract_state_metadata(const json& response,
                                            std::map<int, std::string>& state_ids,
                                            std::map<int, int>& time_stamps) {
    if (!response.contains("result") || !response["result"].is_object()) {
        return false;
    }
    
    const auto& result = response["result"];
    if (!result.contains("states") || !result["states"].is_object()) {
        return false;
    }
    
    const auto& states = result["states"];
    
    for (const auto& [key, state] : states.items()) {
        if (!state.is_object()) {
            continue;
        }
        
        if (state.contains("stateType")) {
            try {
                int state_type = state["stateType"].get<int>();
                
                if (state.contains("stateID") && state["stateID"].is_string()) {
                    state_ids[state_type] = state["stateID"].get<std::string>();
                }
                
                if (state.contains("timeStamp")) {
                    time_stamps[state_type] = state["timeStamp"].get<int>();
                }
            } catch (const std::exception&) {
                continue;
            }
        }
    }
    
    return !state_ids.empty() && !time_stamps.empty();
}

json ObservationApi::get_static_map_data(int map_id) {
    std::string url = "https://static1.bytro.com/fileadmin/mapjson/live/" + 
                      std::to_string(map_id) + ".json";
    
    httplib::Client cli("https://static1.bytro.com");
    cli.set_follow_location(true);
    
    httplib::Headers headers = {
        {"Accept", "application/json, text/javascript, */*; q=0.01"}
    };
    
    std::string path = "/fileadmin/mapjson/live/" + std::to_string(map_id) + ".json";
    auto res = cli.Get(path.c_str(), headers);
    
    if (!res || res->status != 200) {
        throw std::runtime_error("Failed to fetch static map data");
    }
    
    return json::parse(res->body);
}

json ObservationApi::get_cookies() const {
    return cookies_;
}

json ObservationApi::get_headers() const {
    return headers_;
}

File: tools/server_observer_rust/src/observation_api.rs
use crate::account_pool::ProxyConfig;
use crate::hub_interface_wrapper::AuthDetails;
use crate::observation_package::ObservationPackage;
use regex::Regex;
use reqwest::header::{COOKIE, HeaderName, HeaderValue};
use reqwest::{Client, Proxy};
use serde_json::{json, Value};
use sha1::{Digest, Sha1};
use std::collections::HashMap;
use std::time::{Duration, SystemTime, UNIX_EPOCH};
use thiserror::Error;

#[derive(Debug, Clone)]
pub struct HttpResponse {
    pub timeout: bool,
    pub status_code: u16,
    pub data: String,
    pub error_message: String,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum GameServerError {
    Success,
    HttpError,
    ParseError,
    AuthError,
    ServerSwitch,
    ClientVersionMismatch,
    NetworkError,
    UnknownError,
}

#[derive(Debug, Clone)]
pub struct GameServerResult {
    pub error_code: GameServerError,
    pub error_message: String,
    pub raw_response: String,
    pub game_ended: bool,
    pub map_id: String,
}

impl GameServerResult {
    pub fn success(&self) -> bool {
        self.error_code == GameServerError::Success
    }
}

#[derive(Debug, Error)]
pub enum ObservationApiError {
    #[error("http client error: {0}")]
    HttpClient(String),
    #[error("request error: {0}")]
    Request(#[from] reqwest::Error),
    #[error("json parse error: {0}")]
    Json(#[from] serde_json::Error),
}

pub struct ObservationApi {
    game_id: i32,
    player_id: i32,
    auth: AuthDetails,
    request_id: i32,
    client_version: i32,
    game_server_address: String,
    headers: HashMap<String, String>,
    cookies: HashMap<String, String>,
    proxy: ProxyConfig,
    client: Client,
}

impl ObservationApi {
    pub fn new(
        headers: HashMap<String, String>,
        cookies: HashMap<String, String>,
        proxy: ProxyConfig,
        auth_details: AuthDetails,
        game_id: i32,
        game_server_address: String,
        client_version: i32,
    ) -> Result<Self, ObservationApiError> {
        let client = build_http_client(&proxy)?;
        Ok(Self {
            game_id,
            player_id: 0,
            auth: auth_details,
            request_id: 0,
            client_version,
            game_server_address,
            headers,
            cookies,
            proxy,
            client,
        })
    }

    pub async fn request_game_state_async(
        &mut self,
        state_ids: &mut HashMap<String, String>,
        time_stamps: &mut HashMap<String, String>,
    ) -> HttpResponse {
        let include_state_meta = !state_ids.is_empty() && !time_stamps.is_empty();

        let mut action = json!({
            "@c": "ultshared.action.UltUpdateGameStateAction",
            "stateType": 0,
            "stateID": "0",
            "addStateIDsOnSent": include_state_meta,
            "actions": ["java.util.LinkedList", []]
        });

        if include_state_meta {
            action["stateIDs"] = json!(state_ids);
            action["stateIDs"]["@c"] = json!("java.util.HashMap");
            action["tstamps"] = json!(time_stamps);
            action["tstamps"]["@c"] = json!("java.util.HashMap");
        }

        let hash_input = format!("undefined{}", current_time_ms());
        let mut hasher = Sha1::new();
        hasher.update(hash_input.as_bytes());
        let hash_hex = format!("{:x}", hasher.finalize());

        let mut payload = json!({
            "requestID": self.request_id,
            "language": "en",
            "version": self.client_version,
            "tstamp": self.auth.auth_tstamp.to_string(),
            "client": "con-client",
            "hash": hash_hex,
            "sessionTstamp": 0,
            "gameID": self.game_id.to_string(),
            "playerID": self.player_id,
            "siteUserID": self.auth.user_id.to_string(),
            "adminLevel": Value::Null,
            "rights": self.auth.rights,
            "userAuth": self.auth.auth,
            "lastCallDuration": 0
        });
        self.request_id += 1;

        if let Some(map) = action.as_object() {
            for (k, v) in map {
                payload[k] = v.clone();
            }
        }

        let mut req = self
            .client
            .post(&self.game_server_address)
            .header("Accept", "text/plain, */*; q=0.01")
            .header("Accept-Encoding", "gzip, deflate, br")
            .header("Content-Type", "application/json");

        for (k, v) in &self.headers {
            if let (Ok(name), Ok(value)) = (
                HeaderName::from_bytes(k.as_bytes()),
                HeaderValue::from_str(v),
            ) {
                req = req.header(name, value);
            }
        }

        if !self.cookies.is_empty() {
            let cookie_header = self
                .cookies
                .iter()
                .map(|(k, v)| format!("{k}={v}"))
                .collect::<Vec<_>>()
                .join("; ");
            req = req.header(COOKIE, cookie_header);
        }

        let response = req.body(payload.to_string()).send().await;
        let response = match response {
            Ok(resp) => resp,
            Err(err) => {
                return HttpResponse {
                    timeout: err.is_timeout(),
                    status_code: 0,
                    data: String::new(),
                    error_message: err.to_string(),
                };
            }
        };

        let status = response.status().as_u16();
        match response.text().await {
            Ok(data) => HttpResponse {
                timeout: false,
                status_code: status,
                data,
                error_message: String::new(),
            },
            Err(err) => HttpResponse {
                timeout: err.is_timeout(),
                status_code: status,
                data: String::new(),
                error_message: err.to_string(),
            },
        }
    }

    pub fn parse_and_validate_response(
        &mut self,
        response: HttpResponse,
        state_ids: &mut HashMap<String, String>,
        time_stamps: &mut HashMap<String, String>,
    ) -> GameServerResult {
        if response.timeout {
            return GameServerResult {
                error_code: GameServerError::NetworkError,
                error_message: if response.error_message.is_empty() {
                    "Request timed out".to_string()
                } else {
                    response.error_message
                },
                raw_response: String::new(),
                game_ended: false,
                map_id: String::new(),
            };
        }

        if response.status_code != 200 {
            let mut message = format!("HTTP status: {}", response.status_code);
            if !response.error_message.is_empty() {
                message.push_str(" - ");
                message.push_str(&response.error_message);
            }
            return GameServerResult {
                error_code: GameServerError::HttpError,
                error_message: message,
                raw_response: String::new(),
                game_ended: false,
                map_id: String::new(),
            };
        }

        let raw = response.data;
        let game_state: Value = match serde_json::from_str(&raw) {
            Ok(v) => v,
            Err(err) => {
                return GameServerResult {
                    error_code: GameServerError::ParseError,
                    error_message: format!("JSON parse error: {err}"),
                    raw_response: raw,
                    game_ended: false,
                    map_id: String::new(),
                };
            }
        };

        let Some(result_obj) = game_state.get("result").and_then(Value::as_object) else {
            return GameServerResult {
                error_code: GameServerError::UnknownError,
                error_message: "No result object in response".to_string(),
                raw_response: raw,
                game_ended: false,
                map_id: String::new(),
            };
        };

        let Some(result_class) = result_obj.get("@c").and_then(Value::as_str) else {
            return GameServerResult {
                error_code: GameServerError::UnknownError,
                error_message: "No @c field in result".to_string(),
                raw_response: raw,
                game_ended: false,
                map_id: String::new(),
            };
        };

        if result_class == "ultshared.UltAuthentificationException" {
            let mut error_message = format!("Authentication failed: {result_class}");
            if let Some(msg) = result_obj.get("message").and_then(Value::as_str) {
                error_message.push_str(" - ");
                error_message.push_str(msg);
            }
            return GameServerResult {
                error_code: GameServerError::AuthError,
                error_message,
                raw_response: raw,
                game_ended: false,
                map_id: String::new(),
            };
        }

        if result_class == "ultshared.rpc.UltSwitchServerException" {
            let mut error_message = "Server switch required".to_string();
            if let Some(new_server) = result_obj.get("newHostName").and_then(Value::as_str) {
                error_message.push_str(": ");
                error_message.push_str(new_server);
                self.update_server_address(format!("https://{new_server}"));
                if let Ok(client) = build_http_client(&self.proxy) {
                    self.client = client;
                }
            }

            return GameServerResult {
                error_code: GameServerError::ServerSwitch,
                error_message,
                raw_response: raw,
                game_ended: false,
                map_id: String::new(),
            };
        }

        if result_class == "ultshared.UltClientVersionMismatchException" {
            let mut error_message = "Client version mismatch".to_string();
            if let Some(message) = result_obj.get("detailMessage").and_then(Value::as_str) {
                let old_re = Regex::new(r"con-client: #(\d+)").ok();
                let new_re = Regex::new(r"con-client_live\.txt: #(\d+)").ok();

                let old_version = old_re
                    .as_ref()
                    .and_then(|re| re.captures(message))
                    .and_then(|c| c.get(1))
                    .and_then(|m| m.as_str().parse::<i32>().ok());
                let new_version = new_re
                    .as_ref()
                    .and_then(|re| re.captures(message))
                    .and_then(|c| c.get(1))
                    .and_then(|m| m.as_str().parse::<i32>().ok());

                if let (Some(old), Some(new)) = (old_version, new_version) {
                    error_message = format!(
                        "Client version mismatch: old version {old}, new version {new}"
                    );
                    self.client_version = new;
                } else {
                    error_message = format!("Client version mismatch: {message}");
                }
            }

            return GameServerResult {
                error_code: GameServerError::ClientVersionMismatch,
                error_message,
                raw_response: raw,
                game_ended: false,
                map_id: String::new(),
            };
        }

        if result_class != "ultshared.UltAutoGameState" && result_class != "ultshared.UltGameState" {
            let mut message = format!("Unknown error class: {result_class}");
            if let Some(m) = result_obj.get("message").and_then(Value::as_str) {
                message.push_str(" - ");
                message.push_str(m);
            }
            return GameServerResult {
                error_code: GameServerError::UnknownError,
                error_message: message,
                raw_response: raw,
                game_ended: false,
                map_id: String::new(),
            };
        }

        let Some(states) = result_obj.get("states").and_then(Value::as_object) else {
            return GameServerResult {
                error_code: GameServerError::UnknownError,
                error_message: "Game state extraction failed".to_string(),
                raw_response: raw,
                game_ended: false,
                map_id: String::new(),
            };
        };

        let mut game_ended = false;
        let mut map_id = String::new();

        for (key, state_value) in states {
            let Some(state) = state_value.as_object() else {
                continue;
            };

            if let Some(state_id) = state.get("stateID").and_then(value_to_string) {
                state_ids.insert(key.clone(), state_id);
            }
            if let Some(ts) = state.get("timeStamp").and_then(value_to_string) {
                time_stamps.insert(key.clone(), ts);
            }

            if key == "3" {
                if let Some(id) = state
                    .get("map")
                    .and_then(Value::as_object)
                    .and_then(|m| m.get("mapID"))
                    .and_then(value_to_string)
                {
                    map_id = id;
                }
            } else if key == "12" {
                if state
                    .get("gameEnded")
                    .and_then(Value::as_bool)
                    .unwrap_or(false)
                {
                    game_ended = true;
                }
            }
        }

        if state_ids.is_empty() || time_stamps.is_empty() {
            return GameServerResult {
                error_code: GameServerError::UnknownError,
                error_message: "Game state extraction failed".to_string(),
                raw_response: raw,
                game_ended: false,
                map_id: String::new(),
            };
        }

        GameServerResult {
            error_code: GameServerError::Success,
            error_message: String::new(),
            raw_response: raw,
            game_ended,
            map_id,
        }
    }

    pub async fn get_static_map_data(&self, map_id: &str) -> Result<Value, ObservationApiError> {
        let url = format!("https://static1.bytro.com/fileadmin/mapjson/live/{map_id}.json");
        let mut req = self
            .client
            .get(url)
            .header("Accept", "application/json, text/javascript, */*; q=0.01");

        if !self.cookies.is_empty() {
            let cookie_header = self
                .cookies
                .iter()
                .map(|(k, v)| format!("{k}={v}"))
                .collect::<Vec<_>>()
                .join("; ");
            req = req.header(COOKIE, cookie_header);
        }

        let response = req.send().await?;
        if !response.status().is_success() {
            return Err(ObservationApiError::HttpClient(format!(
                "failed static map fetch: status {}",
                response.status()
            )));
        }

        Ok(response.json::<Value>().await?)
    }

    pub fn update_package(&self, pkg: &mut ObservationPackage) {
        pkg.client_version = self.client_version;
        pkg.game_server_address = self.game_server_address.clone();
        pkg.headers = self.headers.clone();
        pkg.cookies = self.cookies.clone();
        pkg.auth = self.auth.clone();
    }

    pub fn update_server_address(&mut self, url: String) {
        self.game_server_address = url;
    }

    pub fn set_proxy(&mut self, proxy: ProxyConfig) -> Result<(), ObservationApiError> {
        self.proxy = proxy;
        self.client = build_http_client(&self.proxy)?;
        Ok(())
    }
}

fn build_http_client(proxy: &ProxyConfig) -> Result<Client, ObservationApiError> {
    let mut builder = Client::builder().timeout(Duration::from_secs(60));
    if proxy.enabled && !proxy.host.is_empty() && proxy.port > 0 {
        let proxy_url = proxy.to_url();
        let req_proxy = Proxy::all(proxy_url)
            .map_err(|err| ObservationApiError::HttpClient(err.to_string()))?;
        builder = builder.proxy(req_proxy);
    }
    builder
        .build()
        .map_err(|err| ObservationApiError::HttpClient(err.to_string()))
}

fn current_time_ms() -> i64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_millis() as i64)
        .unwrap_or_default()
}

fn value_to_string(value: &Value) -> Option<String> {
    match value {
        Value::String(s) => Some(s.clone()),
        Value::Number(n) => Some(n.to_string()),
        Value::Bool(b) => Some(b.to_string()),
        _ => None,
    }
}
